pkgs/development/rocm-modules/5/llvm/stage-3/mlir.nix

```diff
@@ -0,0 +1,69 @@
+{ stdenv
+, callPackage
+, rocmUpdateScript
+, clr
+, vulkan-headers
+, vulkan-loader
+, glslang
+, shaderc
+, lit
+}:
+
+callPackage ../base.nix rec {
+  inherit stdenv rocmUpdateScript;
+  buildDocs = false; # No decent way to hack this to work
+  buildMan = false; # No man pages to build
+  targetName = "mlir";
+  targetDir = targetName;
+  extraNativeBuildInputs = [ clr ];
+
+  extraBuildInputs = [
+    vulkan-headers
+    vulkan-loader
+    glslang
+    shaderc
+  ];
+
+  extraCMakeFlags = [
+    "-DMLIR_INCLUDE_DOCS=ON"
+    "-DMLIR_INCLUDE_TESTS=ON"
+    "-DMLIR_ENABLE_ROCM_RUNNER=ON"
+    "-DMLIR_ENABLE_SPIRV_CPU_RUNNER=ON"
+    "-DMLIR_ENABLE_VULKAN_RUNNER=ON"
+    "-DROCM_TEST_CHIPSET=gfx000" # CPU runner
+  ];
+
+  extraPostPatch = ''
+    chmod +w ../llvm
+    mkdir -p ../llvm/build/bin
+    ln -s ${lit}/bin/lit ../llvm/build/bin/llvm-lit
+
+    # `add_library cannot create target "llvm_gtest" because an imported target with the same name already exists`
+    substituteInPlace CMakeLists.txt \
+      --replace "EXISTS \''${UNITTEST_DIR}/googletest/include/gtest/gtest.h" "FALSE"
+
+    substituteInPlace test/CMakeLists.txt \
+      --replace "FileCheck count not" "" \
+      --replace "list(APPEND MLIR_TEST_DEPENDS mlir_rocm_runtime)" ""
+
+    substituteInPlace lib/ExecutionEngine/CMakeLists.txt \
+      --replace "return()" ""
+
+    # Remove problematic tests
+    rm test/CAPI/execution_engine.c
+    rm test/Target/LLVMIR/llvmir-intrinsics.mlir
+    rm test/Target/LLVMIR/llvmir.mlir
+    rm test/Target/LLVMIR/openmp-llvm.mlir
+    rm test/mlir-cpu-runner/*.mlir
+    rm test/mlir-vulkan-runner/*.mlir
+  '';
+
+  extraPostInstall = ''
+    mkdir -p $out/bin
+    mv bin/mlir-tblgen $out/bin
+  '';
+
+  checkTargets = [ "check-${targetName}" ];
+  requiredSystemFeatures = [ "big-parallel" ];
+  isBroken = true; # `DebugTranslation.cpp:139:10: error: no matching function for call to 'get'`
```

Comment:
It looks like this was fixed in https://github.com/RadeonOpenCompute/llvm-project/commit/f1d1e10ec7e1061bf0b90abbc1e298d9438a5e74, but I'm not sure why the tagged 5.7.0 release is broken.

Reply:
I was able to build this with:

```diff
diff --git a/pkgs/development/rocm-modules/5/llvm/stage-3/fix-debug-translation.patch b/pkgs/development/rocm-modules/5/llvm/stage-3/fix-debug-translation.patch
new file mode 100644
index 000000000000..f319969bf866
--- /dev/null
+++ b/pkgs/development/rocm-modules/5/llvm/stage-3/fix-debug-translation.patch
@@ -0,0 +1,24 @@
+diff --git a/lib/Target/LLVMIR/DebugTranslation.cpp b/lib/Target/LLVMIR/DebugTranslation.cppb
+index f409aa413cf..441558beed5 100644
+--- a/lib/Target/LLVMIR/DebugTranslation.cpp
++++ b/lib/Target/LLVMIR/DebugTranslation.cpp
+@@ -141,7 +141,8 @@ llvm::DIDerivedType *DebugTranslation::translateImpl(DIDerivedTypeAttr attr) {
+       /*File=*/nullptr, /*Line=*/0,
+       /*Scope=*/nullptr, translate(attr.getBaseType()), attr.getSizeInBits(),
+       attr.getAlignInBits(), attr.getOffsetInBits(),
+-      /*DWARFAddressSpace=*/std::nullopt, /*Flags=*/llvm::DINode::FlagZero);
++      /*DWARFAddressSpace=*/std::nullopt, llvm::dwarf::DW_MSPACE_LLVM_none,
++      /*Flags=*/llvm::DINode::FlagZero);
+ }
+ 
+ llvm::DIFile *DebugTranslation::translateImpl(DIFileAttr attr) {
+@@ -172,7 +173,8 @@ DebugTranslation::translateImpl(DILocalVariableAttr attr) {
+       llvmCtx, translate(attr.getScope()), getMDStringOrNull(attr.getName()),
+       translate(attr.getFile()), attr.getLine(), translate(attr.getType()),
+       attr.getArg(),
+-      /*Flags=*/llvm::DINode::FlagZero, attr.getAlignInBits(),
++      /*Flags=*/llvm::DINode::FlagZero, llvm::dwarf::DW_MSPACE_LLVM_none,
++      attr.getAlignInBits(),
+       /*Annotations=*/nullptr);
+ }
+ 
diff --git a/pkgs/development/rocm-modules/5/llvm/stage-3/mlir.nix b/pkgs/development/rocm-modules/5/llvm/stage-3/mlir.nix
index 0e0dcabec60c..1a5924a553c8 100644
--- a/pkgs/development/rocm-modules/5/llvm/stage-3/mlir.nix
+++ b/pkgs/development/rocm-modules/5/llvm/stage-3/mlir.nix
@@ -15,6 +15,11 @@ callPackage ../base.nix rec {
   buildMan = false; # No man pages to build
   targetName = "mlir";
   targetDir = targetName;
+
+  extraPatches = [
+    ./fix-debug-translation.patch
+  ];
+
   extraNativeBuildInputs = [ clr ];
 
   extraBuildInputs = [
@@ -42,19 +47,9 @@ callPackage ../base.nix rec {
     substituteInPlace CMakeLists.txt \
       --replace "EXISTS \''${UNITTEST_DIR}/googletest/include/gtest/gtest.h" "FALSE"
 
-    substituteInPlace test/CMakeLists.txt \
-      --replace "FileCheck count not" "" \
-      --replace "list(APPEND MLIR_TEST_DEPENDS mlir_rocm_runtime)" ""
-
-    substituteInPlace lib/ExecutionEngine/CMakeLists.txt \
-      --replace "return()" ""
-
     # Remove problematic tests
-    rm test/CAPI/execution_engine.c
-    rm test/Target/LLVMIR/llvmir-intrinsics.mlir
-    rm test/Target/LLVMIR/llvmir.mlir
     rm test/Target/LLVMIR/openmp-llvm.mlir
-    rm test/mlir-cpu-runner/*.mlir
+    rm test/mlir-spirv-cpu-runner/*.mlir
     rm test/mlir-vulkan-runner/*.mlir
   '';
 
@@ -65,5 +60,4 @@ callPackage ../base.nix rec {
 
   checkTargets = [ "check-${targetName}" ];
   requiredSystemFeatures = [ "big-parallel" ];
-  isBroken = true; # `DebugTranslation.cpp:139:10: error: no matching function for call to 'get'`
 }
-- 
```

but still very confused why the release version is broken 😅

Reply:
I'm very confused as well. ROCm is a very fast-moving target, guess they didn't notice the issue was in a release version.
Maybe we'll see a patch version addressing it soon lol?

Reply:
Fixed in new commit.